File: templates/problems/06_linked_list/02_merge_two_lists/python/solution.py
from helpers import ListNode, Optional


def solve_merge_two_lists(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    """Merge two sorted linked lists.

    Time: O(n+m), Space: O(1).
    """
    dummy = ListNode()
    curr = dummy
    while list1 is not None and list2 is not None:
        if list1.val <= list2.val:
            curr.next = list1
            list1 = list1.next
        else:
            curr.next = list2
            list2 = list2.next
        curr = curr.next
    if list1 is not None:
        curr.next = list1
    else:
        curr.next = list2
    return dummy.next

File: templates/problems/06_linked_list/02_merge_two_lists/typescript/solution.ts
import { ListNode } from "./helpers";

export function solveMergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const dummy = new ListNode();
    let curr = dummy;
    while (list1 !== null && list2 !== null) {
        if (list1.val <= list2.val) {
            curr.next = list1;
            list1 = list1.next;
        } else {
            curr.next = list2;
            list2 = list2.next;
        }
        curr = curr.next;
    }
    if (list1 !== null) {
        curr.next = list1;
    } else {
        curr.next = list2;
    }
    return dummy.next;
}
